File: model/src/main/java/model/organization/event/NeedsEvent.java
package model.organization.event;

import java.util.Objects;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

import message.M;
import model.organization.entity.Attribute;
import model.organization.entity.Role;
import model.organization.id.UniqueId;
import model.organization.relation.Needs;

import com.google.inject.assistedinject.Assisted;

/**
 * The {@linkplain NeedsEvent} class indicates that there is an update about a {@linkplain Needs} relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
public class NeedsEvent extends AbstractEvent {

	private static final long serialVersionUID = -2368504360543452399L;
	private final UniqueId<Role> roleId;
	private final UniqueId<Attribute> attributeId;
	private transient Integer hashCode = null;
	private transient String toString = null;

	@Inject
	NeedsEvent(@NotNull @Assisted final EventCategory category, @NotNull @Assisted final Needs needs) {
		super(category);
		roleId = needs.getRole().getId();
		attributeId = needs.getAttribute().getId();
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Role}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Role> getRoleId() {
		return roleId;
	}

	/**
	 * Returns a {@linkplain UniqueId} that represents a {@linkplain Attribute}.
	 *
	 * @return a {@linkplain UniqueId}.
	 */
	public UniqueId<Attribute> getAttributeId() {
		return attributeId;
	}

	@Override
	public boolean equals(final Object object) {
		if (object instanceof NeedsEvent) {
			final NeedsEvent event = (NeedsEvent) object;
			return super.equals(event) && getRoleId().equals(event.getRoleId()) && getAttributeId().equals(event.getAttributeId());
		}
		return false;
	}

	@Override
	public int hashCode() {
		if (hashCode == null) {
			hashCode = Objects.hash(getCategory(), getRoleId(), getAttributeId());
		}
		return hashCode;
	}

	@Override
	public String toString() {
		if (toString == null) {
			toString = M.EVENT_WITH_2_IDS.get(super.toString(), getRoleId(), getAttributeId());
		}
		return toString;
	}

}

File: chazm-model/src/main/kotlin/runtimemodels/chazm/model/event/NeedsEvent.kt
package runtimemodels.chazm.model.event

import com.google.inject.assistedinject.Assisted
import runtimemodels.chazm.api.entity.Attribute
import runtimemodels.chazm.api.entity.AttributeId
import runtimemodels.chazm.api.entity.Role
import runtimemodels.chazm.api.entity.RoleId
import runtimemodels.chazm.api.id.UniqueId
import runtimemodels.chazm.api.relation.Needs
import runtimemodels.chazm.model.message.M
import java.util.*
import javax.inject.Inject

/**
 * The [NeedsEvent] class indicates that there is an update about a [Needs] relation.
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
open class NeedsEvent @Inject internal constructor(
    @Assisted category: EventType,
    @Assisted needs: Needs
) : AbstractEvent(category) {
    /**
     * Returns a [UniqueId] that represents a [Role].
     *
     * @return a [UniqueId].
     */
    val roleId: RoleId = needs.role.id
    /**
     * Returns a [UniqueId] that represents a [Attribute].
     *
     * @return a [UniqueId].
     */
    val attributeId: AttributeId = needs.attribute.id

    override fun equals(other: Any?): Boolean {
        if (other is NeedsEvent) {
            return super.equals(other) && roleId == other.roleId && attributeId == other.attributeId
        }
        return false
    }

    override fun hashCode(): Int = Objects.hash(category, roleId, attributeId)

    override fun toString(): String = M.EVENT_WITH_2_IDS[super.toString(), roleId, attributeId]

    companion object {
        private const val serialVersionUID = -2368504360543452399L
    }

}
